### PR TITLE
Remembering previously entered values (auto-fill)

### PR DESCRIPTION
**Pull request by cooperka**
Closes `https://github.com/opendatakit/roadmap/issues/30 TODO link in real PR`

Depends on https://github.com/cooperka/javarosa/pull/1 [TODO: update link in real PR]

Allows form designers to refer to the last-saved instance of this particular form in order to retrieve answers from it. Simply add an external instance to your form:

```xml
<instance id="last-saved" src="jr://instance/last-saved"/>
```

The instance `id` can be anything you want.

Then you can access the external data as usual, for example:

```xml
<setvalue event="xforms-ready" ref="/data/foo" value="instance('last-saved')/data/foo"/>
```

#### What has been done to verify that this works as intended?

Extensive testing with [autoFillTest.xml](https://github.com/cooperka/collect/files/2842612/autoFillTest.xml.txt): filling blank forms, saving forms, filling new blank forms, finalizing, exiting without saving, etc.

#### Why is this the best possible solution? Were any other approaches considered?

See extensive discussion at `https://github.com/opendatakit/roadmap/issues/30 TODO link in real PR`. The result of the discussion there was this approach. It makes use of existing logic to accomplish this feature without needing too much new code.

Collect caches the last-saved instance to the form's media directory `formName-media/last-saved.xml` on each save. This file will stay around even if the original saved instance is lost, which is probably what users would expect. The form's media directory is deleted when the blank form is deleted from the device, which is also probably what users would expect.

When parsing new FormDefs, `lastSavedSrc` will be passed to JavaRosa as either `"jr://file/last-saved.xml"` or `null` if no file exists because the form has never been saved. A `null` value means no instance `src` is provided, meaning the instance will be treated as **internal** instead of **external**. The internal `<instance id="last-saved"/>` has no children and thus will provide empty values for anything referring to it.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Form designers will be able to designate questions as "auto-filled" based on the last-saved value for that particular question. I don't anticipate much regression risk here since the old logic isn't significantly modified.

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

Yes [TODO file issue].

It will also require an update to ODK Build if they want to support this [TODO file issue].

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/tasks/FormLoaderTask.java [hunk 1]
```diff
@@ -124,8 +126,8 @@ protected FECWrapper doInBackground(String... path) {
         final File formXml = new File(formPath);
 
         // set paths to /sdcard/odk/forms/formfilename-media/
-        final String formFileName = formXml.getName().substring(0, formXml.getName().lastIndexOf("."));
-        final File formMediaDir = new File(formXml.getParent(), formFileName + "-media");
+        final String formFileName = FileUtils.getFormFileName(formXml);
+        final File formMediaDir = FileUtils.getFormMediaDir(formXml);
```
Comment (smoyte) [inline collect_app/src/main/java/org/odk/collect/android/tasks/FormLoaderTask.java:130]: Nice refactor.

File: collect_app/src/main/java/org/odk/collect/android/upload/InstanceGoogleSheetsUploader.java [hunk 2]
```diff
@@ -294,8 +296,14 @@ private String uploadMediaFile(Instance instance, String fileName) throws Upload
 
     private TreeElement getInstanceElement(String formFilePath, File instanceFile) throws UploadException {
         FormDef formDef;
+
+        // Get the last-saved instance, if it exists.
+        File formXml = new File(formFilePath);
+        File lastSavedFile = FileUtils.getLastSavedFile(formXml);
+        String lastSavedSrc = lastSavedFile.exists() ? "jr://file/" + LAST_SAVED_FILENAME : null;
```
Comment (smoyte) [inline collect_app/src/main/java/org/odk/collect/android/upload/InstanceGoogleSheetsUploader.java]: This occurs in another file. Should probably be DRYed up?

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 3]
```diff
@@ -400,9 +404,25 @@ public static void deleteAndReport(File file) {
...
-        return pathNoExtension + "-media";
+        return getFormFileName(formFilePath) + "-media";
+    }
+
+    public static File getFormMediaDir(File formXml) {
+        final String formFileName = getFormFileName(formXml);
+        return new File(formXml.getParent(), formFileName + "-media");
+    }
+
+    public static File getLastSavedFile(File formXml) {
```
Comment (smoyte) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java:428]: Since this is only used in tandem with the `String lastSavedSrc = lastSavedFile.exists() ? "jr://file/" + LAST_SAVED_FILENAME : null;`, maybe move it in here?

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 4]
```diff
@@ -400,9 +404,25 @@ public static void deleteAndReport(File file) {
...
+    }
+
+    public static String getFormFileName(String formFilePath) {
+        return formFilePath.substring(0, formFilePath.lastIndexOf('.'));
+    }
+
     public static String constructMediaPath(String formFilePath) {
-        String pathNoExtension = formFilePath.substring(0, formFilePath.lastIndexOf('.'));
-        return pathNoExtension + "-media";
+        return getFormFileName(formFilePath) + "-media";
```
Comment (smoyte) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: Very nice tidy!

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 5]
```diff
@@ -400,9 +404,25 @@ public static void deleteAndReport(File file) {
...
+    }
+
+    public static File getFormMediaDir(File formXml) {
+        final String formFileName = getFormFileName(formXml);
+        return new File(formXml.getParent(), formFileName + "-media");
```
Comment (smoyte) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: Perhaps the `-media` suffix should be a constant? Repeating that string literal seems like a bit of a smell.

**Review by smoyte — COMMENTED** [review]
Nice to get to see some of your code again!

Looks great! I love the simplicity of this approach.

Just a few small noticings.

**Comment by smoyte** [comment]
In the PR description, I would reword `Note: Instead of `last-saved` as an ID, you can use anything you want.` as `The instance ID can be anything you want.`

**Comment by cooperka** [comment]
Thanks @smoyth, I agree with all feedback, I'll polish it up and submit the real thing ~tomorrow.
